llvm/lib/Support/CharSet.cpp

```
@@ -0,0 +1,321 @@
+    return std::error_code();
+  } else if (ConvType == UTF8ToIBM1047) {
+    return ConverterEBCDIC::convertToEBCDIC(Source, Result);
+  }
+  llvm_unreachable("Invalid ConvType!");
+  return std::error_code();
+}
+
+#ifdef HAVE_ICU
+struct UConverterDeleter {
+  void operator()(UConverter *Converter) const {
+    if (Converter)
+      ucnv_close(Converter);
+  }
+};
+using UConverterUniquePtr = std::unique_ptr<UConverter, UConverterDeleter>;
+
+class CharSetConverterICU : public details::CharSetConverterImplBase {
+  UConverterUniquePtr FromConvDesc;
+  UConverterUniquePtr ToConvDesc;
+
+public:
+  CharSetConverterICU(UConverterUniquePtr FromConverter,
+                      UConverterUniquePtr ToConverter)
+      : FromConvDesc(std::move(FromConverter)),
+        ToConvDesc(std::move(ToConverter)) {}
+
+  std::error_code convert(StringRef Source,
+                          SmallVectorImpl<char> &Result) const override;
+};
+
+std::error_code
+CharSetConverterICU::convert(StringRef Source,
+                             SmallVectorImpl<char> &Result) const {
+  // Setup the input in case it has no backing data.
+  size_t InputLength = Source.size();
+  const char *In = InputLength ? const_cast<char *>(Source.data()) : "";
+
+  // Setup the output. We directly write into the SmallVector.
+  size_t Capacity = Result.capacity();
+  size_t OutputLength = Capacity;
+  Result.resize_for_overwrite(Capacity);
+  char *Output = static_cast<char *>(Result.data());
+  UErrorCode EC = U_ZERO_ERROR;
+  do {
+    EC = U_ZERO_ERROR;
+    const char *Input = In;
+
+    Output = InputLength ? static_cast<char *>(Result.data()) : nullptr;
+    ucnv_convertEx(&*ToConvDesc, &*FromConvDesc, &Output, Result.end(), &Input,
+                   In + InputLength, /*pivotStart=*/NULL,
+                   /*pivotSource=*/NULL, /*pivotTarget=*/NULL,
+                   /*pivotLimit=*/NULL, /*reset=*/true,
+                   /*flush=*/true, &EC);
+    if (U_FAILURE(EC)) {
```

For some reason, this is not finding any error with input that ends with an incomplete character, e.g.:
```cpp
static const char EarthUTFExtraPartial[] =
    "\x45\x61\x72\x74\x68\xe5\x9c\xb0\xe7\x90\x83\xe5";
```

---

I'm currently investigating this

---

The reason that the error was not being reported was the following found here https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/ucnv_8h.html:

> When a converter encounters an illegal, irregular, invalid or unmappable character its default behavior is to use a substitution character to replace the bad byte sequence. This behavior can be changed by using [ucnv_setFromUCallBack()](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/ucnv_8h.html#a2975b51ae20ae5302a83cbfd926ebb86) or [ucnv_setToUCallBack()](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/ucnv_8h.html#acf5d877019d10500135f3baa95aa94b4) on the converter. The header [ucnv_err.h](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/ucnv__err_8h.html) defines many other callback actions that can be used instead of a character substitution.

I've added a callback function and added the partial character you added. It correctly reports an error in both iconv and icu now. Thanks for catching this issue!